<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <meta name="author" content="Shing Wan CHOI">
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>LDpred-2 - Basic Tutorial for Polygenic Risk Score Analyses</title>
  <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700" />

  <link rel="stylesheet" href="../css/theme.css" />
  <link rel="stylesheet" href="../css/theme_extra.css" />
  <link href="../css/extra.css" rel="stylesheet" />
  <link href="../css/details.css" rel="stylesheet" />
  
  <script>
    // Current page data
    var mkdocs_page_name = "LDpred-2";
    var mkdocs_page_input_path = "ldpred.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-145068952-1', 'auto');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Basic Tutorial for Polygenic Risk Score Analyses</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../base/">1. QC of Base Data</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../target/">2. QC of Target Data</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../cal_prs/">3. Calculating and analysing PRS</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink/">PLINK</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../prsice/">PRSice-2</a>
                    </li>
                </ul>
                <ul class="current">
                    <li class="toctree-l1 current"><a class="reference internal current" href="./">LDpred-2</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#installing-ldpred-2">Installing LDpred-2</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#required-data">Required Data</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#running-prs-analysis">Running PRS analysis</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#0-prepare-workspace">0. Prepare workspace</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#1-read-in-the-phenotype-and-covariate-files">1. Read in the phenotype and covariate files</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#2-load-in-the-genotype-and-summary-statistic-file-using-bigsnpr">2. Load in the genotype and summary statistic file using bigsnpr</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#3-reformat-data-in-preparation-of-ldpred2">3. Reformat data in preparation of LDpred2</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#4-start-running-ldpred-2">4. Start running LDpred 2</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#5-get-the-final-performance-of-the-ldpred-models">5. Get the final performance of the LDpred models</a>
    </li>
        </ul>
    </li>
    </ul>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../lassosum/">lassosum</a>
                    </li>
                </ul>
                <ul>
                    <li class="toctree-l1"><a class="reference internal" href="../plink_visual/">4. Visualizing PRS Results</a>
                    </li>
                </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Basic Tutorial for Polygenic Risk Score Analyses</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>LDpred-2</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/choishingwan/PRS-Tutorial/edit/master/docs/ldpred.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="background">Background<a class="headerlink" href="#background" title="Permanent link">&para;</a></h1>
<p><a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">LDpred-2</a> is one of the dedicated PRS programs which is an <code>R</code> package that uses a Bayesian approach to polygenic risk scoring.</p>
<h2 id="installing-ldpred-2">Installing LDpred-2<a class="headerlink" href="#installing-ldpred-2" title="Permanent link">&para;</a></h2>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The script used here is based on LDpred 2 implemented under bigsnpr version 1.4.7</p>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For more details, please refer to <a href="https://privefl.github.io/bigsnpr/articles/LDpred2.html">LDpred 2's homepage</a></p>
</div>
<p>You can install <code>LDpred</code> and its dependencies in <code>R</code> with the following command:</p>
<pre class="highlight"><code class="language-R">install.packages("remotes")
library(remotes)
remotes::install_github("https://github.com/privefl/bigsnpr.git")</code></pre>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For mac users, you might need to follow the guide <a href="https://thecoatlessprofessor.com/programming/cpp/openmp-in-r-on-os-x/">here</a> to be able to install LDpred2</p>
</div>
<h2 id="required-data">Required Data<a class="headerlink" href="#required-data" title="Permanent link">&para;</a></h2>
<p>We assume that you have the following files (or you can download it from <a href="https://drive.google.com/file/d/1x_G0Gxk9jFMY-PMqwtg6-vdEyUPp5p5u/view?usp=sharing">here</a>):</p>
<table>
<thead>
<tr>
<th align="center">File Name</th>
<th align="center">Description</th>
</tr>
</thead>
<tbody>
<tr>
<td align="center"><strong>Height.QC.gz</strong></td>
<td align="center">The post-QCed summary statistic</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bed</strong></td>
<td align="center">The genotype file after performing some basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.bim</strong></td>
<td align="center">This file contains the SNPs that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.QC.fam</strong></td>
<td align="center">This file contains the samples that passed the basic filtering</td>
</tr>
<tr>
<td align="center"><strong>EUR.height</strong></td>
<td align="center">This file contains the phenotype of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.cov</strong></td>
<td align="center">This file contains the covariates of the samples</td>
</tr>
<tr>
<td align="center"><strong>EUR.eigenvec</strong></td>
<td align="center">This file contains the PCs of the samples</td>
</tr>
</tbody>
</table>
<h2 id="running-prs-analysis">Running PRS analysis<a class="headerlink" href="#running-prs-analysis" title="Permanent link">&para;</a></h2>
<h3 id="0-prepare-workspace">0. Prepare workspace<a class="headerlink" href="#0-prepare-workspace" title="Permanent link">&para;</a></h3>
<p>On some server, you might need to first use the following code in order to run LDpred with multi-thread</p>
<div class="tabbed-set" data-tabs="1:1"><input checked="checked" id="__tabbed_1_1" name="__tabbed_1" type="radio" /><label for="__tabbed_1_1">prepare workspace</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">options(bigstatsr.check.parallel.blas = FALSE)
options(default.nproc.blas = NULL)</code></pre>
</div>
</div>
<h3 id="1-read-in-the-phenotype-and-covariate-files">1. Read in the phenotype and covariate files<a class="headerlink" href="#1-read-in-the-phenotype-and-covariate-files" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="2:1"><input checked="checked" id="__tabbed_2_1" name="__tabbed_2" type="radio" /><label for="__tabbed_2_1">read in phenotype and covariates</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">library(data.table)
library(magrittr)
phenotype &lt;- fread("EUR.height")
covariate &lt;- fread("EUR.cov")
pcs &lt;- fread("EUR.eigenvec")
# rename columns
colnames(pcs) &lt;- c("FID","IID", paste0("PC",1:6))
# generate required table
pheno &lt;- merge(phenotype, covariate) %&gt;%
    merge(., pcs)</code></pre>
</div>
</div>
<h3 id="2-load-in-the-genotype-and-summary-statistic-file-using-bigsnpr">2. Load in the genotype and summary statistic file using bigsnpr<a class="headerlink" href="#2-load-in-the-genotype-and-summary-statistic-file-using-bigsnpr" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="3:1"><input checked="checked" id="__tabbed_3_1" name="__tabbed_3" type="radio" /><label for="__tabbed_3_1">Load genotype with bigsnpr</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># load the library
library(bigsnpr)
# now preprocess the bed file. Only need to do once for each data set
snp_readBed("EUR.QC.bed")
# now attach the genotype object
obj.bigSNP &lt;- snp_attach("EUR.QC.rds")
# Next, read in the summary statistic file
sumstats &lt;- bigreadr::fread2("Height.QC.gz")</code></pre>
</div>
</div>
<h3 id="3-reformat-data-in-preparation-of-ldpred2">3. Reformat data in preparation of LDpred2<a class="headerlink" href="#3-reformat-data-in-preparation-of-ldpred2" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="4:1"><input checked="checked" id="__tabbed_4_1" name="__tabbed_4" type="radio" /><label for="__tabbed_4_1">data reformating</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># LDpred 2 require the header to follow the exact naming
names(sumstats) &lt;-
    c("chr",
    "pos",
    "rsid",
    "a1",
    "a0",
    "n_eff",
    "beta_se",
    "p",
    "OR",
    "INFO",
    "MAF")
# Transform the OR into log(OR)
sumstats$beta &lt;- log(sumstats$OR)
# extract the SNP information from the genotype
map &lt;- obj.bigSNP$map[-2]
names(map) &lt;- c("chr", "rsid", "pos", "a1", "a0")
# perform SNP matching
info_snp &lt;- snp_match(sumstats, map)
# Rename the data structures
CHR &lt;- map$chr
POS &lt;- map$pos
# get the CM information from 1000 Genome
# will download the 1000G file to the current directory (".")
POS2 &lt;- snp_asGeneticPos(CHR, POS, dir = ".")
# Reformat the phenotype information to match the fam order
fam.order &lt;- as.data.table(obj.bigSNP$fam)
setnames(fam.order,
        c("family.ID", "sample.ID"),
        c("FID", "IID"))
# This will ensure our phenotype vector y is of the same order as the
# family object generated by bigsnpr
y &lt;- pheno[fam.order, on = c("FID", "IID")]
genotype &lt;- obj.bigSNP$genotypes
# Use all sample for test
ind.test &lt;- 1:nrow(genotype)
# Calculate the null R2
# use glm for binary trait 
# (will also need the fmsb package to calculate the pseudo R2)
null.model &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~Sex+", .) %&gt;%
    as.formula %&gt;%
    lm(., data = y) %&gt;%
    summary
null.r2 &lt;- null.model$r.squared
reg.formula &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~PRS+Sex+", .) %&gt;%
    as.formula</code></pre>
</div>
</div>
<h3 id="4-start-running-ldpred-2">4. Start running LDpred 2<a class="headerlink" href="#4-start-running-ldpred-2" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="5:4"><input checked="checked" id="__tabbed_5_1" name="__tabbed_5" type="radio" /><label for="__tabbed_5_1">infinitesimal model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Get maximum amount of cores
NCORES &lt;- nb_cores()
# Start doing analysis on each chromosome
y[, Inf.est := 0]
# add progress bar
pb = txtProgressBar(min = 0, max = 22, initial = 0)
for (chr in 1:22) {
    setTxtProgressBar(pb, chr)
    # extract current chromosome
    chr.idx &lt;- which(info_snp$chr == chr)
    df_beta &lt;- info_snp[chr.idx,
                        c("beta", "beta_se", "n_eff")]
    ind.chr &lt;- info_snp$`_NUM_ID_`[chr.idx]
    # calculate LD
    corr0 &lt;- snp_cor(
        genotype,
        ind.col = ind.chr,
        ncores = NCORES,
        infos.pos = POS2[ind.chr],
        size = 3 / 1000
    )
    corr &lt;- bigsparser::as_SFBM(as(corr0, "dgCMatrix"))
    # Perform LDSC analysis to get h2 estimate
    ldsc &lt;- snp_ldsc2(corr0, df_beta)
    h2_est &lt;- ldsc[["h2"]]
    # Get adjusted beta from infinitesimal model
    beta_inf &lt;- snp_ldpred2_inf(corr, df_beta, h2 = h2_est)
    # Get infinitesimal PRS
    pred_inf &lt;- big_prodVec(genotype,
                            beta_inf,
                            ind.row = ind.test,
                            ind.col = ind.chr)
    # add up the calculated PRS
    y[, Inf.est := Inf.est + pred_inf]
}
print("Completed")</code></pre>
</div>
<input id="__tabbed_5_2" name="__tabbed_5" type="radio" /><label for="__tabbed_5_2">grid model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Get maximum amount of cores
NCORES &lt;- nb_cores()
# Prepare data for grid model
p_seq &lt;- signif(seq_log(1e-4, 1, length.out = 17), 2)
# Start doing analysis on each chromosome
y[, grid := 0]
# add progress bar
pb = txtProgressBar(min = 0, max = 22, initial = 0)
for (chr in 1:22) {
    setTxtProgressBar(pb, chr)
    # extract current chromosome
    chr.idx &lt;- which(info_snp$chr == chr)
    df_beta &lt;- info_snp[chr.idx,
                        c("beta", "beta_se", "n_eff")]
    ind.chr &lt;- info_snp$`_NUM_ID_`[chr.idx]
    # calculate LD
    corr0 &lt;- snp_cor(
        genotype,
        ind.col = ind.chr,
        ncores = NCORES,
        infos.pos = POS2[ind.chr],
        size = 3 / 1000
    )
    corr &lt;- bigsparser::as_SFBM(as(corr0, "dgCMatrix"))
    # Perform LDSC analysis to get h2 estimate
    ldsc &lt;- snp_ldsc2(corr0, df_beta)
    h2_est &lt;- ldsc[["h2"]]
    h2_seq &lt;- round(h2_est * c(0.7, 1, 1.4), 4)
    grid.param &lt;-
        expand.grid(p = p_seq,
                    h2 = h2_seq,
                    sparse = c(FALSE, TRUE))
    # Get adjusted beta from grid model
    beta_grid &lt;-
        snp_ldpred2_grid(corr, df_beta, grid.param, ncores = NCORES)
    # Get the grid PRSs
    pred_grid &lt;- big_prodMat(genotype, beta_grid, ind.col = ind.chr)
    # Find the current "best" PRS
    indep &lt;- copy(y)
    max.grid.r2 &lt;- 0
    max.grid.id &lt;- 0
    for (i in 1:ncol(pred_grid)) {
        indep[, PRS := pred_grid[, i]]
        model &lt;- lm(reg.formula, data = indep) %&gt;%
            summary
        prs.r2 &lt;- model$r.squared - null.r2
        if (max.grid.r2 &lt; prs.r2) {
            max.grid.r2 &lt;- prs.r2
            max.grid.id &lt;- i
        }
    }
    # add up the calculated grid PRS
    y[, grid := grid + pred_grid[, max.grid.id]]
}
print("Completed")</code></pre>
</div>
<input id="__tabbed_5_3" name="__tabbed_5" type="radio" /><label for="__tabbed_5_3">auto model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># Get maximum amount of cores
NCORES &lt;- nb_cores()
# Start doing analysis on each chromosome
y[, auto := 0]
# add progress bar
pb = txtProgressBar(min = 0, max = 22, initial = 0)
for (chr in 1:22) {
    setTxtProgressBar(pb, chr)
    # extract current chromosome
    chr.idx &lt;- which(info_snp$chr == chr)
    df_beta &lt;- info_snp[chr.idx,
                        c("beta", "beta_se", "n_eff")]
    ind.chr &lt;- info_snp$`_NUM_ID_`[chr.idx]
    # calculate LD
    corr0 &lt;- snp_cor(
        genotype,
        ind.col = ind.chr,
        ncores = NCORES,
        infos.pos = POS2[ind.chr],
        size = 3 / 1000
    )
    corr &lt;- bigsparser::as_SFBM(as(corr0, "dgCMatrix"))
    # Perform LDSC analysis to get h2 estimate
    ldsc &lt;- snp_ldsc2(corr0, df_beta)
    h2_est &lt;- ldsc[["h2"]]
    # Get adjusted beta from the auto model
    multi_auto &lt;- snp_ldpred2_auto(
        corr,
        df_beta,
        h2_init = h2_est,
        vec_p_init = seq_log(1e-4, 0.9, length.out = NCORES),
        ncores = NCORES
    )
    beta_auto &lt;- sapply(multi_auto, function(auto)
        auto$beta_est)
    pred_auto &lt;-
        big_prodMat(genotype,
                    beta_auto,
                    ind.row = ind.test,
                    ind.col = ind.chr)
    # scale the PRS generated from AUTO
    pred_scaled &lt;- apply(pred_auto, 2, sd)
    final_beta_auto &lt;-
        rowMeans(beta_auto[,
                        abs(pred_scaled -
                                median(pred_scaled)) &lt;
                            3 * mad(pred_scaled)])
    pred_auto &lt;-
        big_prodVec(genotype,
                    final_beta_auto,
                    ind.row = ind.test,
                    ind.col = ind.chr)
    # add up the calculated auto PRS
    y[, auto := auto + pred_auto]
}
print("Completed")</code></pre>
</div>
<input id="__tabbed_5_4" name="__tabbed_5" type="radio" /><label for="__tabbed_5_4">all model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">
# Get maximum amount of cores
NCORES &lt;- nb_cores()
# Prepare data for grid model
p_seq &lt;- signif(seq_log(1e-4, 1, length.out = 17), 2)
# Start doing analysis on each chromosome
y[, Inf.est := 0]
y[, auto := 0]
y[, grid := 0]
# add progress bar
pb = txtProgressBar(min = 0, max = 22, initial = 0)
for (chr in 1:22) {
    setTxtProgressBar(pb, chr)
    # Extract current chromosome
    chr.idx &lt;- which(info_snp$chr == chr)
    df_beta &lt;- info_snp[chr.idx,
                        c("beta", "beta_se", "n_eff")]
    ind.chr &lt;- info_snp$`_NUM_ID_`[chr.idx]
    # Calculate LD
    corr0 &lt;- snp_cor(
        genotype,
        ind.col = ind.chr,
        ncores = NCORES,
        infos.pos = POS2[ind.chr],
        size = 3 / 1000
    )
    corr &lt;- bigsparser::as_SFBM(as(corr0, "dgCMatrix"))
    # Perform LDSC analysis
    ldsc &lt;- snp_ldsc2(corr0, df_beta)
    h2_est &lt;- ldsc[["h2"]]
    h2_seq &lt;- round(h2_est * c(0.7, 1, 1.4), 4)
    grid.param &lt;-
        expand.grid(p = p_seq,
                    h2 = h2_seq,
                    sparse = c(FALSE, TRUE))
    # Get adjusted beta from infinitesimal model
    beta_inf &lt;- snp_ldpred2_inf(corr, df_beta, h2 = h2_est)
    # Get adjusted beta from grid model
    beta_grid &lt;-
        snp_ldpred2_grid(corr, df_beta, grid.param, ncores = NCORES)
    # Get adjusted beta from the auto model
    multi_auto &lt;- snp_ldpred2_auto(
        corr,
        df_beta,
        h2_init = h2_est,
        vec_p_init = seq_log(1e-4, 0.9, length.out = NCORES),
        ncores = NCORES
    )
    beta_auto &lt;- sapply(multi_auto, function(auto)
        auto$beta_est)
    pred_auto &lt;-
        big_prodMat(genotype,
                    beta_auto,
                    ind.row = ind.test,
                    ind.col = ind.chr)
    # scale the PRS generated from AUTO
    pred_scaled &lt;- apply(pred_auto, 2, sd)
    final_beta_auto &lt;-
        rowMeans(beta_auto[,
                        abs(pred_scaled -
                                median(pred_scaled)) &lt;
                            3 * mad(pred_scaled)])
    # Get the auto PRS
    pred_auto &lt;-
        big_prodVec(genotype,
                    final_beta_auto,
                    ind.row = ind.test,
                    ind.col = ind.chr)
    # Get infinitesimal PRS
    pred_inf &lt;-
        big_prodVec(genotype,
                    beta_inf,
                    ind.row = ind.test,
                    ind.col = ind.chr)
    # Get the grid PRSs
    pred_grid &lt;- big_prodMat(genotype, beta_grid, ind.col = ind.chr)
    # add up the calculated PRS
    y[, Inf.est := Inf.est + pred_inf]
    # for grid, we want to retain the "best" prs only
    indep &lt;- copy(y)
    max.grid.r2 &lt;- 0
    max.grid.id &lt;- 0
    for (i in 1:ncol(pred_grid)) {
        indep[, PRS := pred_grid[, i]]
        model &lt;- lm(reg.formula, data = indep) %&gt;%
            summary
        prs.r2 &lt;- model$r.squared - null.r2
        if (max.grid.r2 &lt; prs.r2) {
            max.grid.r2 &lt;- prs.r2
            max.grid.id &lt;- i
        }
    }
    # add up the calculated grid PRS
    y[, grid := grid + pred_grid[, max.grid.id]]
    # add up the calculated auto PRS
    y[, auto := auto + pred_auto]
}
print("Completed")</code></pre>
</div>
</div>
<h3 id="5-get-the-final-performance-of-the-ldpred-models">5. Get the final performance of the LDpred models<a class="headerlink" href="#5-get-the-final-performance-of-the-ldpred-models" title="Permanent link">&para;</a></h3>
<div class="tabbed-set" data-tabs="6:4"><input checked="checked" id="__tabbed_6_1" name="__tabbed_6" type="radio" /><label for="__tabbed_6_1">infinitesimal model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># there are 3 models: inf, auto and grid
inf.model &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~Inf.est+Sex+", .) %&gt;%
    as.formula %&gt;%
    lm(., data = y) %&gt;%
    summary

result &lt;- data.table(
    infinitesimal = inf.model$r.squared - null.r2,
    null = null.r2
)
print(result)</code></pre>
</div>
<input id="__tabbed_6_2" name="__tabbed_6" type="radio" /><label for="__tabbed_6_2">grid model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">grid.model &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~grid+Sex+", .) %&gt;%
    as.formula %&gt;%
    lm(., data = y) %&gt;%
    summary

result &lt;- data.table(
    grid = grid.model$r.squared - null.r2,
    null = null.r2
)
print(result)</code></pre>
</div>
<input id="__tabbed_6_3" name="__tabbed_6" type="radio" /><label for="__tabbed_6_3">auto model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R">auto.model &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~auto+Sex+", .) %&gt;%
    as.formula %&gt;%
    lm(., data = y) %&gt;%
    summary

result &lt;- data.table(
    infinitesimal = inf.model$r.squared - null.r2,
    grid = grid.model$r.squared - null.r2,
    auto = auto.model$r.squared - null.r2,
    null = null.r2
)
print(result)</code></pre>
</div>
<input id="__tabbed_6_4" name="__tabbed_6" type="radio" /><label for="__tabbed_6_4">all model</label><div class="tabbed-content">
<pre class="highlight"><code class="language-R"># there are 3 models: inf, auto and grid
inf.model &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~Inf.est+Sex+", .) %&gt;%
    as.formula %&gt;%
    lm(., data = y) %&gt;%
    summary
auto.model &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~auto+Sex+", .) %&gt;%
    as.formula %&gt;%
    lm(., data = y) %&gt;%
    summary
grid.model &lt;- paste("PC", 1:6, sep = "", collapse = "+") %&gt;%
    paste0("Height~grid+Sex+", .) %&gt;%
    as.formula %&gt;%
    lm(., data = y) %&gt;%
    summary

result &lt;- data.table(
    infinitesimal = inf.model$r.squared - null.r2,
    grid = grid.model$r.squared - null.r2,
    auto = auto.model$r.squared - null.r2,
    null = null.r2
)
print(result)</code></pre>
</div>
</div>
<details class="note"><summary>How much phenotypic variation does the PRS from each model explain?</summary><p>Infinitesimal = 0.1717342</p>
<p>Grid Model = 0.002945278</p>
<p>Auto Model = 0.2532782</p>
</details>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../lassosum/" class="btn btn-neutral float-right" title="lassosum">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../prsice/" class="btn btn-neutral" title="PRSice-2"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="versions">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/choishingwan/PRS-Tutorial/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../prsice/" style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../lassosum/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js?config=TeX-MML-AM_CHTML" defer></script>
      <script src="../javascripts/details.js" defer></script>
      <script src="../search/main.js" defer></script>
    <script defer>
        window.onload = function () {
            SphinxRtdTheme.Navigation.enable(true);
        };
    </script>

</body>
</html>
